### ISSUE TITLE
Return search results for partial matches

### ISSUE THREAD

**Issue by DoumanAsh** [rust-lang/crates.io#700]
For example take this query: https://crates.io/search?q=irc-channel

It gives you exact match.

But if you attempt to remove last letter then you got nothing at all https://crates.io/search?q=irc-channe

Similar issue with `_`:
https://crates.io/search?q=lazy_static
https://crates.io/search?q=lazy_st

Would it be possible to improve search?

**Comment by carols10cents** [comment]
It's certainly possible!

**Comment by carols10cents** [comment]
I don't think this is only a bug when crates contain `-` or `_`, there's a crate named [acacia](https://crates.io/crates/acacia) but [a search for the partial name acaci](https://crates.io/search?q=acaci) does not return any results.

**Comment by DoumanAsh** [comment]
Oh, i didn't notice this. Then it is not a bug, but enhancement.
Anyway it would be good to have

**Comment by nisargthakkar** [comment]
@carols10cents I believe this issue is arising as a result of postgres `ts_query` and `ts_vector` implementations. According to the code, we do `plainto_tsquery` on the entered query string. This separates the tokens by `&`.
The first search is made on the column containing `tsvector` of all keywords, crate name,description, etc. To search this, the `matches` (`@@`) operator is used.

**Comment by carols10cents** [comment]
@nisargthakkar Can you send a pull request with some test cases that fail before the changes that need to be made and pass after?

**Comment by nisargthakkar** [comment]
@carols10cents I can come up with some test cases but I'm not sure if they'll cover all scenarios as I don't have a good knowledge or any experience working on PostgreSQL

**Comment by carols10cents** [comment]
@nisargthakkar Oh, your comment made it sound like you knew what was going on and what the problem was. What did you mean by your comment then?

**Comment by nisargthakkar** [comment]
@carols10cents I was just going through trying to see what was going on and felt that using a LIKE operator might have fixed it. But later I found that it was using PostgreSQL specifics. So, out of curiosity, I went and read up a bit on what the query builder would churn out given the current code base and what the results might be. Since all the `tsquery` and `tsvector` concept is totally new to me, I might not be the right person to to implement a fix for this issue.
